WIP: refactor Update module to enable testing and add test prototype

yfyf opened the pull request:
This is in an unfinished state (esp. the test part), but pushing this out for early review/feedback.

This PR does the following:
- Refactors the core of `Update` module to use abstract interfaces for its external dependencies (RAUC, Curl/Connman, runtime configuration) rather than direct bindings / values.
- Splits the `Update.run` function into a separate single transition function (`run_step`) and the infinite recursive process (`run`). 
- Exposes an alternative interface to the Update process as a functor over the external dependencies.
- Adds a prototype test "framework" and a sample unit test for the `Update` module.

Some of the code is shifted around/generalized as a side-effect of the changes, but this PR should (hopefully!) not introduce any logical / semantic changes, it only creates new ways for interacting with the code and the sample tests.

This is WIP since several things need to be worked out in more detail:
- The dependency interfaces are just straightforward wrappers around the functions used in the original code, but more thought could be put into this to figure out what exactly are we abstracting over.
- The code split into the `update_deps` module/file is rather ad-hoc
- The test "framework" is now quite basic in terms of what it can test, but my ocaml skills are too limited to figure out how to create mock interfaces neatly[^1].
- Other server modules should probably share (at least) the RAUC abstract interfaces too? 

I also left some TODOs as inline comments. 

@knuton let me know of what you think about this "direction" and specific things that you think need to be done differently. If the general direction seems reasonable, I would attempt to flesh out the test framework and add more complex tests cases. Any ocaml-foo tips are also very welcome.

Apologies for the huge diff, the process was very unlinear, so did not produce structured git commits.

P.S. Is there a specific ocaml style used at Dividat? Maybe we could adopt one of the standard ocamlformat profiles?

[^1]: I want to be able to record every mock call and update the mocked responses at run time. I have an idea how to do this with a simple PPX "decorator", but no idea how to do it otherwise. There's [ocaml-mock](https://github.com/cryptosense/ocaml-mock), but it's very primitive / requires lots of boilreplate.

knuton commented:
Thanks!

Pushing to wrap up some feature work for an app release now, and I have only had the time for a first skim. Some high level comments for now.

> * Refactors the core of `Update` module to use abstract interfaces for its external dependencies (RAUC, Curl/Connman, runtime configuration) rather than direct bindings / values.

Good

> * Splits the `Update.run` function into a separate single transition function (`run_step`) and the infinite recursive process (`run`).

Good

> * Exposes an alternative interface to the Update process as a functor over the external dependencies.

Good, I think

> * Adds a prototype test "framework" and a sample unit test for the `Update` module.

Good, I think

> * The dependency interfaces are just straightforward wrappers around the functions used in the original code, but more thought could be put into this to figure out what exactly are we abstracting over.
> * The code split into the `update_deps` module/file is rather ad-hoc
> * Other server modules should probably share (at least) the RAUC abstract interfaces too?

I think these are related, and also what I was struggling with from a first look. For just this module, the abstractions seem a bit heavy-handed at the moment. My hunch is that this would benefit from thinking about it in a slightly wider scope, as in how could other modules consume this? Even if then possibly only implementing in the context of this one module at first.

yfyf commented:
>
> I think these are related, and also what I was struggling with from a
> first look. For just this module, the abstractions seem a bit heavy-handed
> at the moment. My hunch is that this would benefit from thinking about it
> in a slightly wider scope, as in how could other modules consume this? Even
> if then possibly only implementing in the context of this one module at
> first.
>
Agree, I did not really put much thought into this and it has to be refined
w.r.t. the other modules/components. I will attempt this later this week,
but I definitely want to avoid this becoming a full-blown refactoring of
service composition within the codebase.

However, I pushed this out for early review because I am interested in your
thoughts on the overall mechanics (i.e. interfaces for dependencies +
functors + first-class modules) of this approach. It feels a bit
un-ocaml-ish to me, but then again - I cannot think of a much better
alternative? Been reading discussions
<https://discuss.ocaml.org/t/the-shape-design-problem/7810> (also
<https://discuss.ocaml.org/t/functors-for-dependency-injection/736>, also
<https://mcclurmc.wordpress.com/2012/12/18/ocaml-pattern-easy-functor/>) on
different approaches, but there does not seem to be a consensus on the
"best" way.

> Message ID: ***@***.***>
>

yfyf commented:
@knuton pushed updates that attempt  to tidy-up / generalize things:
- RAUC gets its own "service" interface, added the `mark_good` function to it which would make it immediately re-usable in other modules where it is used.
- Could not think of a useful abstraction that would apply to the Curl wrapper, so essentially left it as is, only moved it to a separate file. It does not seem to be useful outside of the `update` module, as the only other place where proxy resolution is needed is `gui`, but there it is dealing with proxy settings at a much lower level. An alternative could be to create an `UpdateHTTPClient` interface/module, but I would prefer to not introduce new abstractions at this point.
-  Converted the `ConfigInterface` into a plain record, because it's supposed to be just data anyway.
- Added `init _` methods to the interfaces that provide a canonical way to initialize these services / wrappers, to be used in the top-level entrypoint (i.e. `server.ml`)

Also, answering my own question the mechanics, the [Ocaml manual states](https://ocaml.org/manual/5.2/firstclassmodules.html#p:fst-mod-example) that _"a typical use of first-class modules is to select at run-time among several implementations of a signature"_, so I feel it kinda validates this approach.

P.S. I think I also figured out how to elegantly wrap the test implementations into a mock'ed interface, so will update the test prototype soon too.

yfyf commented:
Went all in with splitting the code into modular components and testing all of them independently. Also abandoned the "mock" approach, instead went for stub-implementations of the dependencies.

The changes introduce 3 new abstractions/modules:
- [RaucServiceIntf](https://github.com/yfyf/playos/tree/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/rauc_service.mli) - methods for interacting with RAUC, implementation is just [a functor that wraps the OBus peer reference](https://github.com/yfyf/playos/tree/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/rauc_service.ml#L22), stub for tests is [Mock_rauc](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/tests/mock_rauc.ml) 
- [UpateClientIntf](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/update_client.mli#L9) - interface for obtaining version metadata and downloading bundles, [implementation](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/update_client.ml#L45) is obtained by moving out related functions from the `Update` module. Also a functor, parameterized over the config parameters. Stub for tests is [Mock_update_client](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/tests/mock_update_client.ml). This replaces the kind of useless `CurlProxyInterface` I had earier.
- [UpdateService](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/update.mli#L45) which uses the above two, the interface is essentially the same as in the last changeset.

I also moved all of the runtime / nix-provided system configuration into a single [config](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/config/config.ml) library as a starting point for a more structured approach towards config management. This is necessary to enable runtime test setup, but also sed-template-replacements-as-configuration is kind of a massive hack, so this at least consolidates it into once location and we can think where to go from here.

To run the tests, do
```
CURL_EXECUTABLE_PATH="" dune test -f
```
(The env var / path should later be specified in either the dune config for tests, or, more generally as part of nix-shell setup? Something to be discussed.)  

There are separate tests for `update_client` and for the whole of `update` / `update_service`. Many more should be added for `update_service`, but I actually need specification on what is supposed to happen in various scenarios, it is a bit unclear.

I also littered the code with various `TODO`s and `NOTE`s for potential future improvements / fixes, but I tried to avoid changing any of the semantics, as before.

P.S. Since `curl` is invoked via a sub-process, could we not replace all of the manual-passing-around-of-proxy in the code by simply setting the `HTTPS_PROXY` env variable in the nix-equivalent of `/etc/profile` when [it is setup in `gui.ml`](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/gui.ml#L366-L368)? Unless you maintain separate proxy configs per connman service, but it [does not seem so](https://github.com/dividat/playos/blob/26dc0346d813f4f38cf847e9d0e5134955f8c990/controller/server/update.ml#L117).

knuton commented:
> P.S. Since `curl` is invoked via a sub-process, could we not replace all of the manual-passing-around-of-proxy in the code by simply setting the `HTTPS_PROXY` env variable in the nix-equivalent of `/etc/profile` when [it is setup in `gui.ml`](https://github.com/yfyf/playos/blob/a56b893a52e36e8f4a8097e5b7afa98e93a8df12/controller/server/gui.ml#L366-L368)? Unless you maintain separate proxy configs per connman service, but it [does not seem so](https://github.com/dividat/playos/blob/26dc0346d813f4f38cf847e9d0e5134955f8c990/controller/server/update.ml#L117).

Assuming you are suggesting to just read `HTTPS_PROXY` from an environment variable at service startup once: I think this wouldn't work, because (a) the proxy should be picked up immediately (without restart) when the configuration is changed, and (b) the active proxy might change depending on which network service is the current gateway (which might change without the controller's involvement).

knuton commented on controller/server/update.mli:
```diff
@@ -26,4 +26,30 @@ type state =
   | ReinstallRequired
 [@@deriving sexp]
 
+type sleep_duration = float (* seconds *)
```
I would generally either just use a float or, if defining a dedicated type, leave it abstract in the `.mli`.

```ocaml
type sleep_duration
```

knuton reviewed (commented):
A collection of initial and mostly stylistic comments as I work my way from bottom up.

knuton reviewed (commented):
Nice! I like the little spec language.

Going to send a lightweight spec list next.

yfyf commented on controller/server/update.mli:
```diff
@@ -26,4 +26,14 @@ type state =
+module UpdateService : functor (_ : Update_deps.UpdateServiceDeps) -> sig
+  val run : set_state:(state -> unit) -> state -> unit Lwt.t
+
+  (* private functions used in testing *)
+  module Private : sig
+    val run_step : state -> state Lwt.t
+  end
```
No strong feelings from my side on the matter, so OK, I'll just drop the `Private` sub-module.

yfyf commented on controller/config/config.ml:
```diff
@@ -0,0 +1,23 @@
+    let bundle_name =
+      "@PLAYOS_BUNDLE_NAME@"
+
+    (** If set to empty string, will pick curl from PATH *)
+    let curl_executable_path =
+        Sys.getenv_opt "CURL_EXECUTABLE_PATH" |>
+        Option.value ~default:"/run/current-system/sw/bin/curl"
```
Seems reasonable, will drop the config option entirely.

yfyf commented on controller/server/rauc_service.ml:
```diff
@@ -0,0 +1,50 @@
+module type RaucServiceIntf = sig
```
Kinda agree, but I am wondering whether it makes sense to instead rename it to simply `S` (following the convention in the Ocaml library to denote `S` as the "output signature" of `Make` functors). Then the fully-qualified name is `Rauc_service.S` (`.Intf`), which makes more sense than `Rauc_service.RaucService`?

P.S. Is there any way to configure dune to convert under_scored filenames to CamelCased module names?

knuton replied:
`S` is good :+1: 

(As far as I am aware, the awkwardness of `My_module` is the norm in OCaml.)

yfyf commented on controller/server/update_client.ml:
```diff
@@ -0,0 +1,93 @@
+
+(* NOTE/TODO: potential simplification:
+
+    UpdateClient does not care and need to know about the possibility of a
+    proxy, since we are interfacing with curl via bindings that spawn a
+    subprocess it would be sufficient to set the appropriate `http_proxy` env
+    variable system-wide (at run time) and let curl figure out whether it needs
+    to use a proxy or not itself.
```
After reconsidering, the env variable approach would be messy, but there's a nicer/simpler alternative: write the `proxy` flag value to `curlrc`, which is read on every `curl` invocation (see `curl` man pages on `--config`).

Above you mentioned that:

> the active proxy might change depending on which network service is the current gateway (which might change without the controller's involvement).

These state updates can be managed by listening to `ServicesChanged` / `PropertyChanged` connman signals, which I see are already exposed in the ocaml bindings. I will look into this, but maybe not in the scope of this PR.

yfyf replied:
I realised that the current code is broken already - if the default proxy changes, `UpdateClientConfig.proxy` will never be updated. I started digging and proxy resolution really is all over the place in the code, e.g. the `kiosk` code is [listening to connman proxy updates](https://github.com/dividat/playos/blob/26dc0346d813f4f38cf847e9d0e5134955f8c990/kiosk/kiosk_browser/proxy.py#L135-L150) separately. Consolidating this could simplify a lot of things.

knuton replied:
Yes, global proxy resolution on Linux is a higgledy-piggledy affair. There is no one way to configure a system-wide proxy. We treat connman as the source of truth, but have to query it whenever needed.

When you have KDE or Gnome, they can instead [provide global settings which browsers are in the habit of picking up](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/proxy_config.md). But we'd still have to teach the controller how to get the same settings.

Our conclusion has been that there is no clean and simple way. I'm happy to be proven wrong.

yfyf replied:
Switched back to dynamic proxy lookups from connman in  UpdateClient and removed the comment for now. Let's tackle in a separate task, also keeping in mind other proxy config lookups (e.g. kiosk).

yfyf commented on controller/tests/update_tests.ml:
```diff
@@ -0,0 +1,99 @@
+      StateReached (Installing (_MAGIC_PAT ^ expected_bundle_name next_version));
+      ActionDone
+        ( "bundle was installed and marked as primary",
+          fun () ->
```
Yes, absolutely, that was the original idea - to parameterize the `fun`'s over the test state, but I just didn't get to it. Will attempt to refactor it in this fashion.

yfyf replied:
I pushed an update that resolves this, but I parameterizing over the whole scenario spec, rather than individual scenario specs, because the scenario always uses the same mocks anyway, so no need to pass them each time.

yfyf commented on controller/server/update.ml:
```diff
@@ -106,149 +34,240 @@ type state =
+module UpdateService(Deps : ServiceDeps) = struct
+    open Deps
+
+    let sleep_error_backoff =
+        Lwt_unix.sleep config.error_backoff_duration
+
+    let sleep_update_check =
+        Lwt_unix.sleep config.check_for_updates_interval
```
Yes, since we are not running many tests or "large" quickcheck tests, it's not a big issue to have micro-sleeps. This also allows us to modify these values as needed in one place, rather than having many sleeps with hard-coded values.

In general, I'd rather not get into "mocking" concurrent / IO side-effects, because removing various blocking/IO actions can easily mask implementation bugs in tests, that might actually arise at runtime. It is also non-trivial to get it right.

yfyf commented on controller/tests/update_tests.ml:
```diff
@@ -0,0 +1,99 @@
+  let expected_state_sequence =
+    [
+      UpdateMock (fun () ->
+        Mock_rauc.set_status SystemA {
+            Mock_rauc.some_status with version = current_version
+        };
```
You mean the test case should define the full `Rauc.Slot.status` record, rather than relying on some "base" record (`Rauc_mock.some_status`)? I see what you mean, but the only slot status field that is actually used by the `UpdateService` is the version, so everything else is just noise when it comes to the test specification.

Maybe `Mock_rauc` should just provide a method `set_version : slot -> version -> ()`, instead of having to deal with the whole status?

My general philosophy is that the test should be explicit _only_ about the things that are being tested in it. So if the test is not testing anything about the `status.device` or `status.installed_timestamp`, then why would it specify it? It makes reading/understanding the test harder, because you assume all specified values have some significance.

knuton replied:
> My general philosophy is that the test should be explicit _only_ about the things that are being tested in it. So if the test is not testing anything about the `status.device` or `status.installed_timestamp`, then why would it specify it? It makes reading/understanding the test harder, because you assume all specified values have some significance.

I see that point, but it also goes the other way: How do I know the test only relies on the explicitly specified data, and not other, shared properties? It is easy for a test author to accidentally rely on more than is made explicit, which in turn makes it harder to understand a test's intended assumptions.

yfyf replied:
> How do I know the test only relies on the explicitly specified data, and not other, shared properties?

Well, you almost never know this, since the outcomes can (unintentionally) depend on arbitrary things (anything from developer-machine specific things, e.g. locale, to implementation quirks of the libraries in use). However, by being explicit you are at least saying "my assumption is that specified values / properties `A`, `B`, `C` fully determine the outcome". If this assumption is valid - all good, if not - well then you have a bug either in your implementation or your test spec, since you assumed some inputs are irrelevant, while they actually are.

However, I think you are maybe referring to problems with having many "shared" implicit assumptions in the (helper) test code, which is slightly a different issue.

I'll change it to use the full state, it's fine either way.

yfyf replied:
Updated the test cases [to explicitly define the full RAUC slot status](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/tests/update_tests.ml#L14-L25) - is this how you see it done? It is [kinda repetitive](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/tests/update_tests.ml#L61-L67) and after adding N more test scenarios there will be an itch to generalize, at which point we would kinda go back square one.

knuton commented:
Sat down to write a spec, but instead accidentally refactored the `run` function to separate effectful queries from the version info processing logic: https://github.com/knuton/playos/commits/testing-pure-fun

Even with the added abstractions for effects, I think this kind of reorganization should make testing considerably easier.

yfyf commented on controller/server/update_client.mli:
```diff
@@ -0,0 +1,35 @@
+(* HTTP Client for gettig updates and their metadata from
+   the remote server. *)
+
+(* unparsed semver version *)
+type version = string
+(* local filesystem path *)
+type bundle_path = string
```
Abstract types can be a bit heavy if all you need is e.g. a numeric value for representing seconds. But I think you do need either type aliases or named fun arguments for documentation purposes, otherwise you get interfaces with signatures like `download_bundle: string -> string -> string Lwt.t`, which can be very confusing.

I guess we can decide on a convention of named arguments rather than type aliases?

yfyf commented on controller/server/rauc_service.mli:
```diff
@@ -0,0 +1,26 @@
+(* would suggest moving this to `rauc_service_intf.ml` and 
+   using `include`s (jane-street-style) to avoid having to duplicate
+   it in the .ml files *)
```
I tried experimenting with [dune's `modules_without_implementation` option](https://dune.readthedocs.io/en/stable/reference/dune/library.html), but it seemed to be even more awkward - I could not get them to compose with the rest of the code without defining a separate dune library for each module-without-implementation. Dune docs also say that mli-only modules "are not officially supported by the OCaml compiler", so not sure it's a great idea to use them.

yfyf replied:
After more experimenting, I think it's best to just remain with duplicating the definitions in regular `.ml/.mli` files, unfortunately.

The `_intf` approach does not compose well, leads to issues with `ppx_deriving` and has more issues ([see this discussion for more context](https://discuss.ocaml.org/t/de-duplicating-module-signatures-that-depend-on-abstract-data-types/3826/7).

`ppx_import` could be an alternative, but I don't think the extra pre-processing is helpful, it just makes it harder to debug things when builds subtly break.

I will remove the comments on using the top-down approach.

knuton replied:
Good, makes sense.

yfyf commented:
> refactored the `run` function to separate effectful queries from the version info processing logic

Incorporated these changes, but I went further and made `evaluate_version_info` totally pure, because it does not need to log the error case - [it will be done anyway once the state machine enters the `ErrorGettingVersionInfo` state](https://github.com/dividat/playos/blob/97c12ea7c612ebc824f10638878d1f9ce65b1149/controller/server/update.ml#L150-L155). So now it just returns the next state.

It could be potentially simplified even further by writing higher level helper funs for dealing with semver comparisons RAUC booted/primary slot semantics, but let's leave it for the future.

yfyf commented:
Pushed some semi-final (hopefully!) changes:
- `UpdateClient` now looks up the current proxy dynamically from connman (as it was done originally)
- No more global test state - all mocks/stubs are now object instances that get initialized on each test case, tests are parameterized over a `test_context`. (I tried a module/functor based approach, but it just felt like re-inventing classes and objects/instances rather than using them directly, so I went ahead with objects. It still requires [some boilerplate](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/tests/mock_rauc.ml#L68-L74), but it's somewhat more straightforward in mu opinion).
- some minor improvements / clearing TODOs and dead code
- I noticed [`Config.System`](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/config/config.ml#L4-L13) now partially duplicates the [role of `Info`](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/server/info.ml#L15-L25). I added an `include Config.System` in `info.ml` for now,  not sure it's optimal.

At this point I see two main remaining things:
- More test scenarios for UpdateService
- I kinda want to rename `mock` everywhere into `stub`, because `mock` would imply doing some assertions on how the mock was used (was-method-called, etc), whereas here we just have a test implementation. What do you think?

yfyf commented:
@knuton pushed updates to the earlier tests and a new set of tests that attempt to exhaustively check all the possible (162) combinations as per the spec you provided, [which I distilled into this function](https://github.com/dividat/playos/blob/3770444a79f3ba9e4cfa1593c932c9d82f87ae43/controller/tests/update_test_helpers.ml#L197-L210).

Some cases (well, actually about a third of them) are failing now (see [attachment](https://github.com/user-attachments/files/16850282/tests.log)), because the current implementation does not match the spec. Many are irrelevant and are due to the spec not distinguishing between "doing nothing" and producing an error/warning (and then doing nothing), which can be easily resolved. But some are really questionable, e.g. in test case 90 UpdateService would attempt to update the inactive slot, whereas nothing should be done (according to the spec, but also probably a good idea in practice).

I am using an [intermediate representation of expected outcomes](https://github.com/dividat/playos/blob/3770444a79f3ba9e4cfa1593c932c9d82f87ae43/controller/tests/update_test_helpers.ml#L191-L195) for now to [provide more room for interpreting the results](https://github.com/dividat/playos/blob/3770444a79f3ba9e4cfa1593c932c9d82f87ae43/controller/tests/update_test_helpers.ml#L213-L224).

Will add the side-effect / IO / failure tests next. 

P.S. Alcotest has automagic terminal width detection that leads to truncated output and unreadable test case descriptions. We are running an outdated version of alcotest that does not support the [`ALCOTEST_COLUMNS` env setting](https://github.com/mirage/alcotest/blob/6b628b7d87469cf0c317fd39c51a34dc35f77861/alcotest-help.txt#L80-L83). If you want to get the full output locally, you have to open a "wide" terminal window and run `ALCOTEST_SHOW_ERRORS=1 dune test --no-buffer`.

knuton commented:
This is very interesting

> * 
> * I kinda want to rename `mock` everywhere into `stub`, because `mock` would imply doing some assertions on how the mock was used (was-method-called, etc), whereas here we just have a test implementation. What do you think?

I think you are right, though I admit that I wouldn't have minded calling them "mocks".

> Some cases (well, actually about a third of them) are failing now (see [attachment](https://github.com/user-attachments/files/16850282/tests.log)), because the current implementation does not match the spec. Many are irrelevant and are due to the spec not distinguishing between "doing nothing" and producing an error/warning (and then doing nothing), which can be easily resolved. But some are really questionable, e.g. in test case 90 UpdateService would attempt to update the inactive slot, whereas nothing should be done (according to the spec, but also probably a good idea in practice).

Great. I was indeed struggling with this distinction when drafting a spec, and I think it was part of what led me to think about separating data from interpretation and action.

You did warm me to the goldfish memory design, so maybe it's enough to keep the action separate. What do you think about an interpreter ("free monad") approach to the effects? This may help check the expected action/effect often being "nothing", while still being able to expect a specific interpretation. This may bridge to your intermediate representations.

yfyf commented:
> Running the VM, with `./build vm` and `./result/bin/run-in-vm`, and openning the backdor with `socat` (see command given by the output of _run-in-vm_), I am getting at a high rate the following error message:
> <..>

RAUC is not available when running PlayOS via `run-in-vm`, so this error is kind of expected? It might have been logged differently (or not logged?), but the fact that it is failing to interact with RAUC should not have changed.

yfyf commented on controller/server/rauc_service.ml:
```diff
@@ -0,0 +1,38 @@
+module type S = sig
+    (** [get_status unit] returns current RAUC status *)
+    val get_status : unit -> Rauc.status Lwt.t
+
+    (** [get_booted_slot unit] returns the currently booted slot *)
+    val get_booted_slot : unit -> Rauc.Slot.t Lwt.t
+
+    (** [mark_good slot] marks the slot [slot] as good *)
+    val mark_good : Rauc.Slot.t -> unit Lwt.t
+
+    (** [get_primary unit] returns current primary slot, if any *)
+    val get_primary : unit -> Rauc.Slot.t option Lwt.t
+
+    (** [install source] install the bundle at path [source] *)
+    val install : string -> unit Lwt.t
+end
```
Not sure I understand the question, but it is the usual ocaml duplication of types/signatures defined in interface (`.mli`) files? We discussed potential ways to avoid it with @knuton, but decided there is no elegant way to avoid it.

yfyf commented:
I am converting this back to draft, because while implementing end-to-end tests (PR incoming today), I realized this code will break when deployed, because the nix template variables are being replaced [only in specific modules](https://github.com/dividat/playos/blob/4e49bf185ecfdd0b94a05bd79b4ae303fde100c2/controller/default.nix#L16-L24) and they are moved into `config.ml` in this PR. Will fix it ASAP.

yfyf commented:
> Yes, but the problem is the super high rate (multiple times each second), making the backdoor when developping useless, and flooding the logs. In `main`, the error message shows up, but at a slow rate (every x seconds).

Ah, I see, strange - I will look into it!

yfyf commented:
> > Yes, but the problem is the super high rate (multiple times each second), making the backdoor when developping useless, and flooding the logs. In `main`, the error message shows up, but at a slow rate (every x seconds).
> 
> Ah, I see, strange - I will look into it!

@guyonvarch should be fixed now, see [the last commit](https://github.com/dividat/playos/pull/173/commits/6a27dda16a663ed18d94da62dfeab1ece9d3636b). This was quite an ugly bug, thank you for pointing it out!

yfyf commented on controller/server/update_client.ml:
```diff
@@ -0,0 +1,86 @@
+    >|= Option.map (Connman.Service.Proxy.to_uri ~include_userinfo:true)
+
+let build_module connman =
+  (* TODO: this could take only `unit` as an argument by just getting
+     the connman reference like this:
+  let%lwt connman = Connman.Manager.connect () in *)
```
Are you asking what implications are there if we recreate the connman connection in `update_client.ml`? I would say "none", but I wanted to stay as close to the original architecture as possible and to do further refactoring in a separate PR. I think it's best to simply remove the TODO for now, it is not a "must have".

yfyf replied:
Removed in 1b51d4b

yfyf commented on controller/server/update.ml:
```diff
@@ -104,151 +41,230 @@ type state =
+            return (slot_primary, slot_booted, vsn_resp)
+        ) in
+        (match resp with
+            | Ok (slot_p, slot_b, vsns) ->
```
It's the `version_info` struct, which contains the latest, booted and inactive version*s*. Renamed to `version_info` in 7310791

yfyf commented on controller/server/update.ml:
```diff
@@ -104,151 +41,230 @@ type state =
+end
+
+let default_config : config = {
+    error_backoff_duration = 30.0;
+    check_for_updates_interval = (1. *. 60. *. 60.)
```
This is already specified [in the interface](https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/server/update.mli#L29-L37), is there any point repeating it here?

yfyf commented on controller/tests/update_test_helpers.ml:
```diff
@@ -0,0 +1,366 @@
+
+let state_formatter out inp = Format.fprintf out "%s" (statefmt inp)
+
+let t_state =
```
The `t` stands for testable, because this defines an instance of `Alcotest.testable` for `state`. I could rename it to `testable_state` if that's helpful.

yfyf commented on controller/tests/update_test_helpers.ml:
```diff
@@ -0,0 +1,366 @@
+    let state_eq expected actual =
+        (expected == actual) || (
+            str_match_with_magic_pat
+                (* Using string repr is a horrible hack, but avoids having to
+                   pattern match on every variant in the state ADT *)
```
> Why would we have to match on every variant in the ADT, could you provide an example?

Because we need to define equality for `state`, which would require defining equality for all the cases, including e.g. `UpToDate of version_info`.

yfyf replied:
> Also, I don’t understand at this point where the magic pattern is created, is it in Update.sexp_of_state?

Do you mean it's unclear how it's used? Here are a few places:
- https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/tests/update_tests.ml#L65
- https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/tests/update_tests.ml#L77
- https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/tests/update_tests.ml#L12-L14

yfyf replied:
Could be renamed to `_MAGIC_WILDCARD_PAT` or just `_WILDCARD_PAT` without the magic :)

yfyf commented on controller/tests/update_test_helpers.ml:
```diff
@@ -0,0 +1,366 @@
+    Alcotest.testable state_formatter state_eq
+
+
+let interp_spec (state : Update.state) (spec : scenario_spec) =
```
Renamed to `interpret_spec` in 5c84e52

yfyf commented on controller/server/update_client.ml:
```diff
@@ -0,0 +1,86 @@
+module UpdateClient (DepsI: UpdateClientDeps) = struct
+    let get_proxy = DepsI.get_proxy
+    let download_dir = DepsI.download_dir
+    let base_url = ensure_trailing_slash DepsI.base_url
```
Renamed in aed4928

yfyf commented on controller/tests/update_client_tests.ml:
Added a top-level comment in c5f4482

yfyf commented:
> I am WIP to move some code around / create modules. I’ll create a PR on your branch when ready.
> 
> https://github.com/yfyf/playos/pull/1

@guyonvarch incorporated your changes together with some additional minor renaming/shuffling as discussed.

Now what remains is to rebase and resolve the `ppx_protocol_conv_jsonm` conflict, which I am onto now.

yfyf commented:
@guyonvarch rebased on top of main after fixing the global namespace conflicts with #206
